clients/src/main/java/org/apache/kafka/common/network/Selector.java

```diff
@@ -237,12 +239,17 @@ public void close() {
      * @param send The request to send
      */
     public void send(Send send) {
-        KafkaChannel channel = channelOrFail(send.destination());
-        try {
-            channel.setSend(send);
-        } catch (CancelledKeyException e) {
-            this.failedSends.add(send.destination());
-            close(channel);
+        String connectionId = send.destination();
+        if (closingChannels.containsKey(connectionId))
+            this.failedSends.add(connectionId);
```

Comment:
Hmm, it seems that failedSends is never used in the caller to Selector. Let's say the producer sends 2 outstanding produce requests with ack=1 and then closes the socket. A broker gets the first request, processes it and tries to send a response. The send fails and is added to failedSends. However, since SocketServer never checks failedSends, it doesn't know the send fails or the channel is disconnected. So, this socket will never be closed by the broker?